diff -rupN gromacs-5.1-orig/src/gromacs/gmxlib/nonbonded/nonbonded.c gromacs-5.1-hck/src/gromacs/gmxlib/nonbonded/nonbonded.c
--- gromacs-5.1-orig/src/gromacs/gmxlib/nonbonded/nonbonded.c	2015-06-22 15:17:54.000000000 +0200
+++ gromacs-5.1-hck/src/gromacs/gmxlib/nonbonded/nonbonded.c	2018-04-17 12:15:46.960758191 +0200
@@ -241,6 +241,7 @@ gmx_nonbonded_set_kernel_pointers(FILE *
     vdw_mod  = eintmod_names[nl->ivdwmod];
     geom     = gmx_nblist_geometry_names[nl->igeometry];
 
+/* chk
     if (nl->type == GMX_NBLIST_INTERACTION_ADRESS)
     {
         nl->kernelptr_vf       = (void *) gmx_nb_generic_adress_kernel;
@@ -248,6 +249,7 @@ gmx_nonbonded_set_kernel_pointers(FILE *
         nl->simd_padding_width = 1;
         return;
     }
+*/

diff -rupN gromacs-5.1-orig/src/gromacs/mdlib/adress.c gromacs-5.1-hck/src/gromacs/mdlib/adress.c
--- gromacs-5.1-orig/src/gromacs/mdlib/adress.c	2015-06-22 15:17:54.000000000 +0200
+++ gromacs-5.1-hck/src/gromacs/mdlib/adress.c	2018-04-17 12:14:21.588290415 +0200
@@ -110,8 +110,8 @@ adress_weight(rvec                 x,
     /* hybrid region */
     else
     {
-        tmp = cos((dl-adressr)*M_PI/2/adressw);
-        return tmp*tmp;
+        /*tmp = cos((dl-adressr)*M_PI/2/adressw);*/
+        return 0.5;
     }
 }
 
diff -rupN gromacs-5.1-orig/src/gromacs/mdlib/ns.c gromacs-5.1-hck/src/gromacs/mdlib/ns.c
--- gromacs-5.1-orig/src/gromacs/mdlib/ns.c	2015-06-22 15:17:54.000000000 +0200
+++ gromacs-5.1-hck/src/gromacs/mdlib/ns.c	2018-04-17 12:14:03.368190548 +0200
@@ -264,10 +264,11 @@ void init_neighbor_list(FILE *log, t_for
     for (i = 0; i < fr->nnblists; i++)
     {
         nbl = &(fr->nblists[i]);
-
+/* chk */
         if ((fr->adress_type != eAdressOff) && (i >= fr->nnblists/2))
         {
-            type = GMX_NBLIST_INTERACTION_ADRESS;
+	/*  type = GMX_NBLIST_INTERACTION_ADRESS; */
+            type = GMX_NBLIST_INTERACTION_STANDARD;
         }
         init_nblist(log, &nbl->nlist_sr[eNL_VDWQQ], &nbl->nlist_lr[eNL_VDWQQ],
                     maxsr, maxlr, ivdw, ivdwmod, ielec, ielecmod, igeometry_def, type, bElecAndVdwSwitchDiffers);
@@ -601,11 +602,14 @@ put_in_list_at(gmx_bool              bHa
     int          *cginfo;
     int          *type, *typeB;
     real         *charge, *chargeB;
+/*    real         *wf; */
     real          qi, qiB, qq, rlj;
     gmx_bool      bFreeEnergy, bFree, bFreeJ, bNotEx, *bPert;
     gmx_bool      bDoVdW_i, bDoCoul_i, bDoCoul_i_sol;
+    gmx_bool      b_hybrid;
     int           iwater, jwater;
     t_nblist     *nlist;
+/*    gmx_bool      bEnergyGroupCG; */
 
     /* Copy some pointers */
     cginfo  = fr->cginfo;
@@ -614,6 +618,7 @@ put_in_list_at(gmx_bool              bHa
     type    = md->typeA;
     typeB   = md->typeB;
     bPert   = md->bPerturbed;
+/*    wf      = md->wf; */
 
     /* Get atom range */
     i0     = index[icg];
@@ -622,10 +627,12 @@ put_in_list_at(gmx_bool              bHa
     /* Get the i charge group info */
     igid   = GET_CGINFO_GID(cginfo[icg]);
 
+/*    bEnergyGroupCG = !egp_explicit(fr, igid); */
     iwater = (solvent_opt != esolNO) ? GET_CGINFO_SOLOPT(cginfo[icg]) : esolNO;
 
     bFreeEnergy = FALSE;
-    if (md->nPerturbed)
+/*    if (md->nPerturbed && (fr->adress_type == eAdressOff)) */
+    if (md->nPerturbed ) 
     {
         /* Check if any of the particles involved are perturbed.
          * If not we can do the cheaper normal put_in_list
@@ -683,6 +690,7 @@ put_in_list_at(gmx_bool              bHa
     }
 
     if (!bFreeEnergy)
+/*    if (!bFreeEnergy ||  (fr->adress_type != eAdressOff)) */
     {
         if (iwater != esolNO)
         {
@@ -846,8 +854,12 @@ put_in_list_at(gmx_bool              bHa
                 bDoVdW_i  = (bDoVdW  && bHaveVdW[type[i_atom]]);
                 bDoCoul_i = (bDoCoul && qi != 0);
 
+
                 if (bDoVdW_i || bDoCoul_i)
                 {
+				/* chk 
+                    bEnergyGroupCG = !egp_explicit(fr, igid); */
+
                     /* Loop over the j charge groups */
                     for (j = 0; (j < nj); j++)
                     {
@@ -867,7 +879,18 @@ put_in_list_at(gmx_bool              bHa
                         /* Finally loop over the atoms in the j-charge group */
                         for (jj = jj0; jj < jj1; jj++)
                         {
+
                             bNotEx = NOTEXCL(bExcl, i, jj);
+				/* chk 
+
+
+                        if ( ( !bEnergyGroupCG && ( wf[i_atom] <= GMX_REAL_EPS || wf[jj] <= GMX_REAL_EPS ) ) || 
+			     ( ( bEnergyGroupCG ) && ( wf[i_atom] >= GMX_REAL_EPS && wf[jj] >= GMX_REAL_EPS ) ) 
+			     )
+
+                            {
+                                continue;
+                            } */
 
                             if (bNotEx)
                             {
@@ -984,6 +1007,10 @@ put_in_list_at(gmx_bool              bHa
                     jj1 = index[jcg+1];
                     /* Finally loop over the atoms in the j-charge group */
                     bFree = bPert[i_atom];
+
+		    /* chk 
+                    bEnergyGroupCG = !egp_explicit(fr, igid); */
+
                     for (jj = jj0; (jj < jj1); jj++)
                     {
                         bFreeJ = bFree || bPert[jj];
@@ -994,6 +1021,16 @@ put_in_list_at(gmx_bool              bHa
                         {
                             bNotEx = NOTEXCL(bExcl, i, jj);
 
+			    /* chk 
+
+                        if ( ( !bEnergyGroupCG && ( wf[i_atom] <= GMX_REAL_EPS || wf[jj] <= GMX_REAL_EPS ) ) || 
+			     ( ( bEnergyGroupCG ) && ( wf[i_atom] >= GMX_REAL_EPS && wf[jj] >= GMX_REAL_EPS ) ) 
+			     )
+
+                            {
+                                continue;
+                            } */
+
                             if (bNotEx)
                             {
                                 if (bFreeJ)
@@ -1250,6 +1287,18 @@ put_in_list_adress(gmx_bool
                      * b_hybrid=true are placed into the _adress neighbour lists and
                      * processed by the generic AdResS kernel.
                      */
+      /* change 7.11.2017 chk*/     
+                   /* if ( fr->adress_type != eAdressOff ) 
+		       { */
+                        if ( ( !bEnergyGroupCG && ( wf[i_atom] <= GMX_REAL_EPS || wf[jj] <= GMX_REAL_EPS ) ) || 
+			     ( ( bEnergyGroupCG ) && ( wf[i_atom] >= GMX_REAL_EPS && wf[jj] >= GMX_REAL_EPS ) ) 
+			     )
+
+                            {
+                                continue;
+                            } 
+                    /*    } */
+/*    old version from normal GC-AdResS before october 7                 
                     if ( (bEnergyGroupCG &&
                           wf[i_atom] >= 1-GMX_REAL_EPS && wf[jj] >= 1-GMX_REAL_EPS ) ||
                          ( !bEnergyGroupCG && wf[jj] <= GMX_REAL_EPS ) )
@@ -1259,6 +1308,7 @@ put_in_list_adress(gmx_bool
 
                     b_hybrid = !((wf[i_atom] >= 1-GMX_REAL_EPS && wf[jj] >= 1-GMX_REAL_EPS) ||
                                  (wf[i_atom] <= GMX_REAL_EPS && wf[jj] <= GMX_REAL_EPS));
+*/
 
                     if (bNotEx)
                     {
@@ -1266,28 +1316,15 @@ put_in_list_adress(gmx_bool
                         {
                             if (charge[jj] != 0)
                             {
-                                if (!b_hybrid)
-                                {
-                                    add_j_to_nblist(coul, jj, bLR);
-                                }
-                                else
-                                {
-                                    add_j_to_nblist(coul_adress, jj, bLR);
-                                }
+				    /* chk: removed the !b_hybrid if loops */
+                               add_j_to_nblist(coul, jj, bLR);
                             }
                         }
                         else if (!bDoCoul_i)
                         {
                             if (bHaveVdW[type[jj]])
                             {
-                                if (!b_hybrid)
-                                {
                                     add_j_to_nblist(vdw, jj, bLR);
-                                }
-                                else
-                                {
-                                    add_j_to_nblist(vdw_adress, jj, bLR);
-                                }
                             }
                         }
                         else
@@ -1296,38 +1333,16 @@ put_in_list_adress(gmx_bool
                             {
                                 if (charge[jj] != 0)
                                 {
-                                    if (!b_hybrid)
-                                    {
                                         add_j_to_nblist(vdwc, jj, bLR);
-                                    }
-                                    else
-                                    {
-                                        add_j_to_nblist(vdwc_adress, jj, bLR);
-                                    }
                                 }
                                 else
                                 {
-                                    if (!b_hybrid)
-                                    {
                                         add_j_to_nblist(vdw, jj, bLR);
-                                    }
-                                    else
-                                    {
-                                        add_j_to_nblist(vdw_adress, jj, bLR);
-                                    }
-
                                 }
                             }
                             else if (charge[jj] != 0)
                             {
-                                if (!b_hybrid)
-                                {
                                     add_j_to_nblist(coul, jj, bLR);
-                                }
-                                else
-                                {
-                                    add_j_to_nblist(coul_adress, jj, bLR);
-                                }
 
                             }
                         }
@@ -2671,8 +2686,10 @@ int search_neighbours(FILE *log, t_force
     rvec     box_size, grid_x0, grid_x1;
     int      i, j, m, ngid;
     real     min_size, grid_dens;
+    real     b_hybrid;
     int      nsearch;
     gmx_bool     bGrid;
+    gmx_bool     bEnergyGroupCG;
     char     *ptr;
     gmx_bool     *i_egp_flags;
     int      cg_start, cg_end, start, end;
@@ -2773,7 +2790,7 @@ int search_neighbours(FILE *log, t_force
         fill_grid(NULL, ns->grid, fr->hcg, fr->hcg-1, fr->hcg, fr->cg_cm);
     }
     debug_gmx();
-
+/* chk */
     if (fr->adress_type == eAdressOff)
     {
         if (!fr->ns.bCGlist)
