diff -rupN gromacs-5.1-orig/src/gromacs/mdlib/update.cpp gromacs-5.1-hck/src/gromacs/mdlib/update.cpp
--- gromacs-5.1-orig/src/gromacs/mdlib/update.cpp	2015-06-22 15:17:54.000000000 +0200
+++ gromacs-5.1-hck/src/gromacs/mdlib/update.cpp	2018-04-17 12:14:10.620230299 +0200
@@ -623,10 +623,23 @@ static void do_update_sd1(gmx_stochd_t *
             {
                 if ((ptype[n] != eptVSite) && (ptype[n] != eptShell) && !nFreeze[gf][d])
                 {
-                    real sd_V, vn;
+//                    real sd_V, vn;
+		    real sd_V, vn, fn, fc, fc1;
+                   
+                    sd_V         = ism*sig[gt].V*rnd[d];
+
+                    fc           = 10000.;
+                    fn           = f[n][d];
+                    if (fabs(fn)>fc) 
+                       { 
+			  printf("SD (I) force-cap %e\n", fn);
+			  fc1=fc;
+                          fn = fc1*fn/fabs(fn);   
+                       }
 
                     sd_V         = ism*sig[gt].V*rnd[d];
-                    vn           = v[n][d] + (invmass[n]*f[n][d] + accel[ga][d])*dt;
+                    vn           = v[n][d] + (invmass[n]*fn + accel[ga][d])*dt;
+//                    vn           = v[n][d] + (invmass[n]*f[n][d] + accel[ga][d])*dt;
                     v[n][d]      = vn*sdc[gt].em + sd_V;
                     /* Here we include half of the friction+noise
                      * update of v into the integration of x.
@@ -666,7 +679,20 @@ static void do_update_sd1(gmx_stochd_t *
                 {
                     if ((ptype[n] != eptVSite) && (ptype[n] != eptShell) && !nFreeze[gf][d])
                     {
-                        v[n][d]      = v[n][d] + (im*f[n][d] + accel[ga][d])*dt;
+
+                    real fn, fc, fc1;
+       
+                    fc           = 10000.;
+                    fn           = f[n][d];
+                    if (fabs(fn)>fc) 
+                       { 
+			  printf("SD (II) force-cap %e\n", fn);
+			  fc1=fc;
+                          fn = fc1*fn/fabs(fn);   
+                       }
+
+                        v[n][d]      = v[n][d] + (im*fn + accel[ga][d])*dt;
+//                        v[n][d]      = v[n][d] + (im*f[n][d] + accel[ga][d])*dt;
                         xprime[n][d] = x[n][d] +  v[n][d]*dt;
                     }
                     else
